In [55]:
!pip install mlxtend

In [56]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [57]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/7105/Cross-selling models/Customer Preference Survey (Responses) - Form Responses 1.csv')

In [58]:
df.head(5)

Timestamp playstation5  ...        ยาดม ไฟส่องหน้าไลฟ์สด
0  10/17/2021 17:43:38   ไม่เคยซื้อ  ...  ไม่เคยซื้อ       ไม่เคยซื้อ
1  10/17/2021 17:43:41   ไม่เคยซื้อ  ...  ไม่เคยซื้อ       ไม่เคยซื้อ
2  10/17/2021 17:43:46   ไม่เคยซื้อ  ...     เคยซื้อ          เคยซื้อ
3  10/17/2021 17:43:49   ไม่เคยซื้อ  ...  ไม่เคยซื้อ       ไม่เคยซื้อ
4  10/17/2021 17:43:55   ไม่เคยซื้อ  ...     เคยซื้อ       ไม่เคยซื้อ

[5 rows x 42 columns]

In [59]:
missing = df.isnull().sum()
print(missing)

Timestamp                        0
playstation5                     0
เครื่องทำขนมปัง                  0
Ergonomic Wrist Rest             0
เครื่องอบผ้า                     0
เครื่องชงกาแฟแคปซูล              0
เก้าอี้ LA-Z-Boy                 0
เครื่องให้อาหารสัตว์อัตโนมัติ    0
บัตตาเลี่ยน                      1
แก้วเก็บความเย็น                 0
ลู่วิ่งออกกำลังกาย               0
Kindle                           0
เครื่องซักผ้า                    0
Bluetooth Speaker                0
ห้องน้ำแมวอัตโนมัติ              0
PS5                              0
ทรายแมว                          0
ลำโพง pixel                      0
Logitech Mx Master 3 Mouse       0
ตุ๊กตา ty                        0
น้ำพุแมว                         0
Robot ดูดฝุ่น                    0
Mechanical keyboard              1
Nintendo switch                  1
หนังสือ python                   0
gaming chair                     1
Deskmat                          0
Dew - ไฟโรเซ่                    0
เทียนหอม jo malone  

In [60]:
hot_encoded_df=df.fillna(0)

In [61]:
hot_encoded_df = hot_encoded_df.replace(['ไม่','ไม่เคยซื้อ','ไม่เคย','ไม่เคยซือ'],0).replace(['เคย','เคยซื้อ'],1)
hot_encoded_df.head()

Timestamp  playstation5  ...  ยาดม  ไฟส่องหน้าไลฟ์สด
0  10/17/2021 17:43:38             0  ...     0                 0
1  10/17/2021 17:43:41             0  ...     0                 0
2  10/17/2021 17:43:46             0  ...     1                 1
3  10/17/2021 17:43:49             0  ...     0                 0
4  10/17/2021 17:43:55             0  ...     1                 0

[5 rows x 42 columns]

In [62]:
hot_encoded_df = hot_encoded_df.drop(columns='Timestamp')
hot_encoded_df.head()

playstation5  เครื่องทำขนมปัง  ...  ยาดม  ไฟส่องหน้าไลฟ์สด
0             0                0  ...     0                 0
1             0                0  ...     0                 0
2             0                0  ...     1                 1
3             0                0  ...     0                 0
4             0                0  ...     1                 0

[5 rows x 41 columns]

In [63]:
missing = hot_encoded_df.isnull().sum()
print(missing)

playstation5                     0
เครื่องทำขนมปัง                  0
Ergonomic Wrist Rest             0
เครื่องอบผ้า                     0
เครื่องชงกาแฟแคปซูล              0
เก้าอี้ LA-Z-Boy                 0
เครื่องให้อาหารสัตว์อัตโนมัติ    0
บัตตาเลี่ยน                      0
แก้วเก็บความเย็น                 0
ลู่วิ่งออกกำลังกาย               0
Kindle                           0
เครื่องซักผ้า                    0
Bluetooth Speaker                0
ห้องน้ำแมวอัตโนมัติ              0
PS5                              0
ทรายแมว                          0
ลำโพง pixel                      0
Logitech Mx Master 3 Mouse       0
ตุ๊กตา ty                        0
น้ำพุแมว                         0
Robot ดูดฝุ่น                    0
Mechanical keyboard              0
Nintendo switch                  0
หนังสือ python                   0
gaming chair                     0
Deskmat                          0
Dew - ไฟโรเซ่                    0
เทียนหอม jo malone               0
กระติกน้ำ 2 ลิตร    

In [64]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
hot_encoded_df = hot_encoded_df.applymap(encode_units)

In [65]:
import numpy as np
import matplotlib.pyplot as plt

In [66]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(hot_encoded_df, min_support=0.1, use_colnames=True)

In [67]:
hot_encoded_df.head(5)

playstation5  เครื่องทำขนมปัง  ...  ยาดม  ไฟส่องหน้าไลฟ์สด
0             0                0  ...     0                 0
1             0                0  ...     0                 0
2             0                0  ...     1                 1
3             0                0  ...     0                 0
4             0                0  ...     1                 0

[5 rows x 41 columns]

In [68]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
hot_encoded_df = hot_encoded_df.applymap(encode_units)

In [69]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(hot_encoded_df, min_support=0.1, use_colnames=True)

In [70]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1).sort_values(by='lift', ascending=False)
rules.head(20)

antecedents  ... conviction
588873  (Salmon Sashimi, เครื่องซักผ้า, ลู่วิ่งออกกำลั...  ...        inf
769791  (Salmon Sashimi, External Harddisk, ไฟแต่งห้อง...  ...        inf
589674  (airpods , หม้อทอดไร้น้ํามัน, พลาสเตอร์บรรเทาป...  ...   5.347826
731358  (Bluetooth Speaker, หม้อทอดไร้น้ํามัน, พลาสเตอ...  ...   5.347826
1440                          (ทรายแมว, แก้วเก็บความเย็น)  ...   5.347826
695024  (Salmon Sashimi, เครื่องซักผ้า, ไฟแต่งห้องมินิ...  ...        inf
590009                (เครื่องซักผ้า, ลู่วิ่งออกกำลังกาย)  ...        inf
695133  (Salmon Sashimi, เครื่องซักผ้า, ไฟแต่งห้องมินิ...  ...        inf
769788  (Salmon Sashimi, External Harddisk, เครื่องซัก...  ...        inf
722031  (Salmon Sashimi, External Harddisk, ไฟแต่งห้อง...  ...        inf
723149  (Salmon Sashimi, เครื่องซักผ้า, ลู่วิ่งออกกำลั...  ...        inf
386035  (airpods , หม้อทอดไร้น้ํามัน, พลาสเตอร์บรรเทาป...  ...   5.347826
722946  (Salmon Sashimi, หม้อทอดไร้น้ํามัน, ยาดม, พลาส...  ...   5.347826
590711  (Salmon Sashimi, ลู่วิ่งออกกำลังกาย, Bluetooth...  ...        inf
386434  (เครื่องซักผ้า, ลู่วิ่งออกกำลังกาย, External H...  ...        inf
742060  (หม้อทอดไร้น้ํามัน, Robot ดูดฝุ่น, ยาดม, Mecha...  ...   5.347826
723145  (Salmon Sashimi, เครื่องซักผ้า, ลู่วิ่งออกกำลั...  ...        inf
590567  (Salmon Sashimi, External Harddisk, ยาดม, หนัง...  ...        inf
581533  (หม้อทอดไร้น้ํามัน, Robot ดูดฝุ่น, ยาดม, Mecha...  ...   5.347826
190031  (airpods , หม้อทอดไร้น้ํามัน, พลาสเตอร์บรรเทาป...  ...   5.347826

[20 rows x 9 columns]

In [71]:
rules[ (rules['lift'] >= 1) &
       (rules['confidence'] >= 0.5) ]

antecedents  ... conviction
588873  (Salmon Sashimi, เครื่องซักผ้า, ลู่วิ่งออกกำลั...  ...        inf
769791  (Salmon Sashimi, External Harddisk, ไฟแต่งห้อง...  ...        inf
589674  (airpods , หม้อทอดไร้น้ํามัน, พลาสเตอร์บรรเทาป...  ...   5.347826
731358  (Bluetooth Speaker, หม้อทอดไร้น้ํามัน, พลาสเตอ...  ...   5.347826
1440                          (ทรายแมว, แก้วเก็บความเย็น)  ...   5.347826
...                                                   ...  ...        ...
26692        (หม้อทอดไร้น้ํามัน, External Harddisk, ยาดม)  ...   1.000000
26704                 (airpods , External Harddisk, ยาดม)  ...   1.000000
170519  (แก้วเก็บความเย็น, External Harddisk, Bluetoot...  ...   1.000000
27564                  (Salmon Sashimi, กระติกน้ำ 2 ลิตร)  ...   1.000000
28061        (Salmon Sashimi, พลาสเตอร์บรรเทาปวด ตราเสือ)  ...   1.000000

[473580 rows x 9 columns]

In [72]:
item_item_matrix = pd.DataFrame(index=hot_encoded_df.columns,columns=hot_encoded_df.columns)

In [73]:
from scipy.spatial.distance import cosine

for i in range(0,len(item_item_matrix.columns)) :
    # Loop through the columns for each column
    for j in range(0,len(item_item_matrix.columns)) :
      # Fill in placeholder with cosine similarities
      item_item_matrix.iloc[i,j] = 1 -cosine(hot_encoded_df.iloc[:,i],hot_encoded_df.iloc[:,j])

In [74]:
links = item_item_matrix.rename_axis('related item', 
                                     axis='columns').stack().reset_index()
links.columns = ['item', 'related item','value']

In [75]:
links_filtered=links.loc[ (links['value'] > 0.1) & 
                         (links['item'] != links['related item']) ]

In [76]:
links_filtered.sort_values(by='value', ascending=False).head(10)

item       related item     value
14         playstation5                PS5         1
574                 PS5       playstation5         1
634             ทรายแมว           น้ำพุแมว  0.912871
794            น้ำพุแมว            ทรายแมว  0.912871
1392     Salmon Sashimi               ยาดม  0.883191
1632               ยาดม     Salmon Sashimi  0.883191
531   Bluetooth Speaker               ยาดม  0.834622
1611               ยาดม  Bluetooth Speaker  0.834622
1635               ยาดม  External Harddisk   0.82674
1515  External Harddisk               ยาดม   0.82674

## summary

ProductRecommendation

จาก basket row 769791 มี cofident 1.0	 and lift 7.666667

antecedents : Salmon Sashimi, External Harddisk, ไฟแต่งห้อง...	

consequents : หม้อทอดไร้น้ํามัน, Robot ดูดฝุ่น, ยาดม, Mecha...

เมื่อลูกค้ามีการซื้อ ไฟแต่งห้อง จะแนะนำ Mechanical keyboard เพื่อให้มีความคลาสสิคเข้ากับไฟแต่งห้อง และแนะนำ Robot ดูดฝุ่น เพื่อทำความสะอาดในห้อง
